### DISCUSSION TITLE
Qwen3-TTS 0.6B CustomVoice 的 ncnn 移植：第一次提交与严格自回归一致性验证

### DISCUSSION THREAD

**Discussion by soso23412** [Tencent/ncnn#6862] [Show and tell]
# Qwen3-TTS 0.6B CustomVoice 的 ncnn 移植：第一次提交与严格自回归一致性验证

这是我针对 [Tencent/ncnn#6791](https://github.com/Tencent/ncnn/issues/6791)
完成的第一次正式成果提交。本文记录实现方案、转换流程、验证方法和对抗性审查中发现的问题，
欢迎大家审阅和提出改进建议。

- 项目仓库：<https://github.com/soso23412/ncnn_llm>
- Windows/Linux CI：<https://github.com/soso23412/ncnn_llm/actions/runs/30239106471>

## 我完成了什么

我将 `Qwen/Qwen3-TTS-12Hz-0.6B-CustomVoice` 通过 pnnx 转换为 ncnn
模型，并实现了从文本到最终波形的 C++20/ncnn 推理程序。部署后的运行时不依赖 Python、
PyTorch、Transformers 或 Qwen3-TTS Python 包。

固定使用英文文本 `Hello, welcome to Qwen text to speech.`、说话人 `Ryan`
和语言 `English` 后，我得到以下结果：

| 测试场景 | 用途 | Codec 一致性 | PCM16 一致性 |
| --- | --- | ---: | ---: |
| 2 帧 | 跨平台冒烟测试 | 32/32 完全一致 | 3,840 个采样点，最大误差 1 LSB |
| 25 帧 | 自回归回归测试 | 400/400 完全一致 | 48,000 个采样点，最大误差 1 LSB |
| 第 54 帧遇到 EOS | 生产停止行为 | 864/864 完全一致 | 103,680 个采样点，最大误差 1 LSB |
| 64 帧 | 忽略 EOS 的数值压力测试 | 1,024/1,024 完全一致 | 仅比较 codes |
| 128 帧 | 忽略 EOS 的数值压力测试 | 2,048/2,048 完全一致 | 仅比较 codes |

其中 64 帧和 128 帧测试会故意忽略 codec EOS，用于检查长序列数值累积、位置和缓存问题。
这两组结果不是有效音频，因此运行时只允许将 `--ignore-eos` 与 `--codes-only`
组合使用。

## 可复现约束

为了避免“同一个模型名称、不同运行结果”的问题，我分别固定了模型和参考实现：

- 模型：`Qwen/Qwen3-TTS-12Hz-0.6B-CustomVoice`
- 模型 revision：`85e237c12c027371202489a0ec509ded67b5e4b5`
- Qwen3-TTS 源码 revision：`022e286b98fbec7e1e916cb940cdf532cd9f488e`
- 参考 dtype：float32
- 标准参考设备：CPU
- Attention 实现：eager
- 生成方式：greedy
- TF32：关闭

`models/qwen3-tts/checkpoint.json` 记录了所有必要 checkpoint 文件的大小和
SHA-256。转换后的 `conversion.json` 会记录转换参数、输入信息以及每个 ncnn
资产的哈希。转换和验证过程不接受 `main` 这类可移动分支作为 revision。

我在对抗性复测中发现，PyTorch CPU 与 CUDA 会在接近并列的 logits 上选择不同的
greedy token。第一次分歧出现在第 6 帧、codebook 11：

- CPU 选择 token `1628`，top-1/top-2 margin 为 `0.0189285`
- CUDA 选择 token `315`，top-1/top-2 margin 为 `0.0063896`

ncnn 与 CPU 参考完全一致，因此我将 CPU 设备作为严格一致性测试向量的一部分，
而不是把 CPU/CUDA 当作可互换的性能选项。完整 top-5 决策信息保存在
`models/qwen3-tts/evidence/pytorch/cpu-vs-cuda-greedy-divergence.json`。

## 运行时拆分

我将推理链路拆成以下部分，并为每部分固定输入输出张量契约：

1. C++ frontend 负责文本分词、CustomVoice prompt 构造，以及 text/codec
   embedding 图执行。
2. Talker prefill 处理完整 prompt 并初始化自回归状态。
3. Talker decode 逐帧产生主 codec token。
4. Code Predictor 为每帧产生其余 15 个 codebook token。
5. RVQ frontend 和 Speech Decoder 将合法 codec 帧转换为 24 kHz 单声道音频。

Code Predictor 最终保留了 15 个固定序列长度的 ncnn 图。这个方案比共享 KV cache
原型占用更多包空间，但能够保持严格的 CPU 参考决策一致性。我曾测试更快的共享 KV
原型，但它在第 10 帧、codebook 7 发生分歧，因此没有把该优化合入本次提交。

另一个关键问题是 codec EOS。Talker 的 EOS token 是控制 token，不是可以送入音频
解码器的 code。生产模式现在会在 EOS 帧之前停止，并把实际生成帧数而不是请求的最大
帧数传给 Speech Decoder。在保留的 128 帧上限测试中，模型会自然地在第 54 帧产生
EOS。

## 模型转换

先校验固定版本的官方 checkpoint：

```bash
python models/qwen3-tts/download_checkpoint.py verify \
  --manifest models/qwen3-tts/checkpoint.json \
  --root /path/to/Qwen3-TTS-12Hz-0.6B-CustomVoice
```

然后运行支持断点续转的转换程序：

```bash
python models/qwen3-tts/convert.py \
  --checkpoint /path/to/Qwen3-TTS-12Hz-0.6B-CustomVoice \
  --output /path/to/qwen3-tts-ncnn \
  --work /path/to/qwen3-tts-work \
  --pnnx /path/to/pnnx \
  --device cpu \
  --frames 25
```

转换程序会显式提供所有 pnnx 输入形状，要求生成 FP32 图，检查应当共享的权重文件是否
完全一致，并拒绝仍包含 `pnnx.*` 或 `Tensor.*` 运行时层的模型。长时间转换中断后可用
`--resume` 复用已完成阶段。

## 构建

使用已经安装的 ncnn 源码目录：

```bash
cmake -S . -B build -G Ninja \
  -DNCNN_SOURCE_DIR=/path/to/ncnn \
  -DCMAKE_BUILD_TYPE=Release
cmake --build build --parallel
ctest --test-dir build --output-on-failure
```

仓库中的 GitHub Actions 会在 `ubuntu-latest` 与 `windows-latest` 上执行相同的
configure、build、test 和 install 流程。本次提交的两个矩阵任务均已通过。

## 运行

```bash
build/qwen3_tts_main \
  --model /path/to/qwen3-tts-ncnn/model.json \
  --frames 128 \
  --text-file models/qwen3-tts/testdata/prompt.txt \
  --speaker Ryan \
  --language English \
  --out output.wav \
  --codes output-codes-i32.bin \
  --threads 8
```

`--frames` 表示生成上限，正常合成可能在 codec EOS 处提前结束。数值诊断时可以使用
`--codes-only`；额外的 `--ignore-eos` 参数只有在 codes-only 模式下才会被接受。

## 一致性验证

比较工具要求 codec token 逐元素完全相等。如果出现分歧，它会报告第一个不同 token
所在的帧和 codebook、完全一致的前缀长度，以及 16 个 codebook 各自的错误数量。
提供音频时，采样率和采样点数必须一致；仅在最终 float 转 PCM16 的边界允许最多一个
整数 LSB 的误差。

```bash
python models/qwen3-tts/compare.py \
  --reference-codes models/qwen3-tts/testdata/pytorch-cpu-exact-25f-codes-i32.bin \
  --reference-wav models/qwen3-tts/testdata/pytorch-cpu-exact-25f.wav \
  --ncnn-codes models/qwen3-tts/testdata/ncnn-windows-exact-25f-codes-i32.bin \
  --ncnn-wav models/qwen3-tts/testdata/ncnn-windows-exact-25f.wav
```

机器可读报告和运行日志位于 `models/qwen3-tts/evidence/`。模型权重单独托管，不会
提交到 Git 仓库。

## 这次移植中的几点经验

- 两帧样例只能证明图连接正确，不能证明长序列自回归稳定。
- 仅写“PyTorch 参考”不够；当 logits 接近并列时，CPU 与 CUDA 可能进入不同分支。
- EOS 属于模型协议的一部分，控制 token 不能作为普通音频 code 送入解码器。
- 缓存优化不能只比较 teacher-forced 张量误差，还必须比较端到端 greedy 生成结果。
- 失败报告应保留 top candidates 和 margin，便于区分数值分支变化与 mask、position、
  cache、shape 或 codebook 顺序错误。

更完整的通过标准和失败处理方式记录在 `docs/parity-contract.md`。这是我的第一次正式
提交，欢迎大家审阅实现、复现结果并提出进一步优化建议。
